### PR TITLE
Adding instrumentation hooks for request and resonses

### PR DESCRIPTION
**Pull request by bmalinconico**
Adds:

1) Retries
2) Sleep & Jitter w/ 429 respect
3) Hooks for logging to DD

See paired MyT PR.

### CODE REVIEW

**Comment by bmalinconico** [comment]
@nrowegt there are a few good candidates for improvements in this gem, testing being the most critical one. I'd prefer not to do that in this PR to limit the scope of the changes.

File: lib/ttdrest/concerns/base.rb [hunk 1]
```diff
@@ -20,16 +25,67 @@ def authenticate(options = {})
...
+
+          check_response(response)
+        rescue RecoverableHttpError
+          raise if retries >= ERROR_RETRIES
+          retries += 1
+
+          if sleep_time = parse_header_retry
+            sleep(sleep_time)
+          else
+            sleep((1.25 ** reties) * 1 - (0.3 * rand))
```
Comment (1BrandonCummings) [inline lib/ttdrest/concerns/base.rb]: reties

**Comment by bmalinconico** [comment]
@1BrandonCummings typo fixed. Testing will be a pain, since there are currently almost no tests... I can poke around.

File: lib/ttdrest/concerns/base.rb [hunk 2]
```diff
@@ -20,16 +25,67 @@ def authenticate(options = {})
...
+        response && (
+          response.code == 429 || (500..599).include?(response.code)
+        )
+      end
+
+      def parse_header_retry(response)
+        return nil unless response['Retry-After']
+
+        if response['Retry-After'] =~ /^\d+$/
+          sleep(response['Retry-After'].to_i)
```
Comment (stephenreid) [inline lib/ttdrest/concerns/base.rb]: questioning the api doc, is this retry after utc time or a number of seconds, this is a number of seconds.

File: lib/ttdrest/concerns/base.rb [hunk 3]
```diff
@@ -20,16 +25,67 @@ def authenticate(options = {})
...
+
+          check_response(response)
+        rescue RecoverableHttpError
+          raise if retries >= ERROR_RETRIES
+          retries += 1
+
+          if sleep_time = parse_header_retry
+            sleep(sleep_time)
+          else
+            sleep((1.25 ** retries) * 1 - (0.3 * rand))
```
Comment (stephenreid) [inline lib/ttdrest/concerns/base.rb]: jitter and backoff

File: lib/ttdrest/concerns/base.rb [hunk 4]
```diff
@@ -20,16 +25,67 @@ def authenticate(options = {})
       def check_response(response)
         if response.code.eql?("403")
```
Comment (stephenreid) [inline lib/ttdrest/concerns/base.rb]: Will this be a string or a symbol?

Should we add other bad response codes, like DNS lookup errors here?

Comment (bmalinconico) [reply lib/ttdrest/concerns/base.rb]: I did confirm it is a string.

Do we know what the class of errors for connection errors are? `SocketError`?

File: lib/ttdrest/concerns/base.rb (hunk 2 above)
Comment (bmalinconico) [reply lib/ttdrest/concerns/base.rb]: Yes the docs say a # of seconds backoff, or a date string.

**Comment by bmalinconico** [comment]
Specs added

**Pull request merged**